python/mxnet/symbol/symbol.py

```diff
@@ -1437,6 +1437,54 @@ def _gen_atomic_symbol(self):
         return Symbol(handle)
 
 
+    def optimize_for(self, backend, args=None, **kwargs):
```

Comment:
I agree with you on having support for passing shape/type/stype dicts just like simple_bind. I will add that support in a later PR. Since we accept kwargs now, it would not be a breaking API change.

I understand your concern about having a more explicit api, but I prefer keeping it this way for the following reason. Users are expected to know what arguments are required by the backend. The backend errors out if the required arguments are not provided whether or not we have an explicit flag. At this point, it's not a choice for users to not infer shape/type/stype when the backend requires it, or vice versa.